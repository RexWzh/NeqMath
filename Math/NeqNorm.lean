import Mathlib

open Real

local macro_rules | `($x / $y)   => `(HDiv.hDiv ($x : ℝ) ($y : ℝ))

/-
This lean file is created for expression normalization
-/

theorem zero_one : 0 + 1 = 1 := by simp

theorem pow_mul_zero (a : ℝ) (b : ℕ) : 0 * a ^ b = 0 := by
  field_simp

theorem inv_reduce (a : ℝ) : a ^ (-(1:ℤ)) = 1 / a := by
  norm_cast; field_simp;


/-
Handle the sqrt reduction
-/
theorem sqrt_reduce (a : ℝ) (h : a ≥ 0): sqrt a ^ 2 = a := by
  apply sq_sqrt h

theorem sqrt_reduce1 (a b : ℝ) (c : ℕ) (ha : a ≥ 0) (hb : b ≠ 0) (hc : b = c) : (a ^ (b⁻¹)) ^ c = a := by
  suffices h : (a ^ (b⁻¹)) ^ c = (a ^ (b⁻¹)) ^ b by
    subst hc
    simp_all only [ge_iff_le, ne_eq, Nat.cast_eq_zero, rpow_natCast]
    rw [← rpow_natCast, ← rpow_mul ha]
    simp_all only [isUnit_iff_ne_zero, ne_eq, Nat.cast_eq_zero, not_false_eq_true, IsUnit.inv_mul_cancel, rpow_one]
  rw [hc]
  norm_cast

theorem sqrt_reduce2 (a c : ℝ) (b : ℕ) (ha : a ≥ 0) (hb : c ≠ 0) (hc : b = c) : (a ^ b) ^ c⁻¹ = a := by
  suffices h : (a ^ b) ^ c⁻¹ = (a ^ c) ^ c⁻¹ by
    subst hc
    simp_all only [ge_iff_le, ne_eq, Nat.cast_eq_zero, rpow_natCast]
    rw [← rpow_natCast, ← rpow_mul ha]
    simp_all only [isUnit_iff_ne_zero, ne_eq, Nat.cast_eq_zero, not_false_eq_true, IsUnit.mul_inv_cancel, rpow_one]
  rw [<-hc]
  norm_cast

/-
Handle the fraction reduction
-/
theorem frac_reduce (a b : ℝ) (h : b ≥ 0) (h' : a ≤ 0) : a / b ≤ 0 := by
  have h : 1 / b ≥ 0 := by positivity
  have h : a * (1 / b) ≤ 0 := by nlinarith
  rw [show a / b = a * (1 / b) by field_simp]
  trivial

theorem frac_reduce' (a b : ℝ) (h : a ≥ 0) (h' : b ≤ 0) : a / b ≤ 0 := by
  have h : 1 / b ≤ 0 := by simp_all only [ge_iff_le, one_div, inv_nonpos]
  have h : a * (1 / b) ≤ 0 := by nlinarith
  rw [show a / b = a * (1 / b) by field_simp]
  trivial

theorem frac_reduce1 (a b c : ℝ) : a / c * b = a * b / c := by
  field_simp [*]

theorem frac_reduce2 (a b: ℝ) (h : a ≠ 0) : 1 / a ^ 2 * b = b / a ^ 2 := by
  field_simp [*]


theorem factor_reduce (a b : ℝ) (h : a ≥ 0) (h' : b ≤ 0) : a * b ≤ 0 := by
  nlinarith

theorem factor_reduce' (a b : ℝ) (h : b ≥ 0) (h' : a ≤ 0) : a * b ≤ 0 := by
  nlinarith

theorem factor_reduce_neg (a b : ℝ) (h : -a ≥ 0) (h' : -b ≤ 0) : a * b ≤ 0 := by
  nlinarith

theorem factor_reduce_neg' (a b : ℝ) (h : b ≥ 0) (h' : a ≤ 0) : a * b ≤ 0 := by
  nlinarith


/-
Rearrange
-/

theorem move_left (left right : ℝ) (h : left - right ≤ 0) : left ≤ right := by nlinarith

theorem move_right (left right : ℝ) (h : 0 ≤ right - left) : left ≤ right := by nlinarith

theorem rearrange (left right : ℝ) (h : left ≤ right) : left - right ≤ 0 := by nlinarith


/-
Variable cancalation lemmas
-/

theorem add6_sub_cancel3 (a b c d e f : ℝ) : a + b + c + d + e + f - c = a + b + d + e + f := by
  abel

theorem add6_sub_cancel4 (a b c d e f : ℝ) : a + b + c + d + e + f - d = a + b + c + e + f := by
  abel

theorem add6_sub_cancel5 (a b c d e f : ℝ) : a + b + c + d + e + f - e = a + b + c + d + f := by
  abel

/-
Variable cancalation lemmas
-/
theorem square_root_of_4 : (4 : ℝ) = 2 ^ 2:= by
  norm_num

theorem cubic_root_of_8 : (8 : ℝ) = 2 ^ 3:= by
  norm_num

theorem cubic_root_of_27 : (27 : ℝ) = 3 ^ 3 := by
  norm_num
