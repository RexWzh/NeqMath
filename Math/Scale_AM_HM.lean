import Mathlib

open Real

local macro_rules | `($x / $y)   => `(HDiv.hDiv ($x : ℝ) ($y : ℝ))

-- ========================== AM-HM inequalities ==========================

/-
  The basic format of the AM-HM inequality for 2 variables
-/
theorem AM_HM_2vars (u v : ℝ) (hu : u > 0) (hv : v > 0) : 2 / (1 / u + 1 / v) ≤ (u + v) / 2 := by
  rw [<- mul_le_mul_right (a := (1 / u + 1 / v)) (b := 2 / (1 / u + 1 / v)) (c := (u + v) / 2)];
  rw [show 2 / (1 / u + 1 / v) * (1 / u + 1 / v) = 2 by field_simp [*]]
  let f : ℕ → ℝ | 0 => sqrt (1 / u) | 1 => sqrt (1 / v) | _ => 0
  let g : ℕ → ℝ | 0 => sqrt (u) | 1 => sqrt (v) | _ => 0
  let s : Finset ℕ := {0, 1}
  have h := Finset.sum_mul_sq_le_sq_mul_sq s f g
  simp [f, s] at h
  simp [g, s] at h
  (repeat rw [sq_sqrt] at h)
  rw [<- mul_le_mul_right (a := 2) (b := 2) (c:= (u + v) / 2 * (1 / u + 1 / v))]
  convert h using 1
  field_simp; norm_num;
  simp; linarith;
  all_goals positivity

/-
  The div form of the AM-HM inequality for 2 variables
-/
theorem AM_HM_div_left_2vars (u v k l right : ℝ) (hu : u > 0) (hv : v > 0) (hk : k > 0) (h : k * (1 / u + 1 / v) / 4 + l ≤ right) : k * (1 / (u + v)) + l ≤ right := by
  suffices 1 / (u + v) ≤ (1 / u + 1 / v) / 4 by nlinarith
  have h := AM_HM_2vars (1 / u) (1 / v)
  field_simp [*] at h
  rw [<- mul_le_mul_right (a := 2) (b := 1 / (u + v)) (c := (1 / u + 1 / v) / 4)]
  convert h using 1
  norm_num; field_simp;
  field_simp; linarith;
  positivity

theorem AM_HM_div_right_2vars (u v k l left : ℝ) (hu : u > 0) (hv : v > 0) (hk : k > 0) (h : left ≤ k * (4 / (u + v)) + l) :
  left ≤ k * (1 / u + 1 / v) + l := by
  suffices 4 / (u + v) ≤ (1 / u + 1 / v) by nlinarith
  have h := AM_HM_2vars (1 / u) (1 / v)
  field_simp [*] at h
  rw [<- mul_le_mul_right (a := 1 / 2) (b := 4 / (u + v)) (c := (1 / u + 1 / v))]
  convert h using 1
  norm_num; field_simp; linarith
  field_simp; linarith;

/-
  The basic form of the AM-HM inequality for 3 variables
-/
theorem AM_HM_3vars (u v w : ℝ) (hu : u > 0) (hv : v > 0) (hw : w > 0) : 3 / (1 / u + 1 / v + 1 / w) ≤ (u + v + w) / 3 := by
  rw [<- mul_le_mul_right (a := (1 / u + 1 / v + 1 / w)) (b := 3 / (1 / u + 1 / v + 1 / w)) (c := (u + v + w) / 3)];
  rw [show 3 / (1 / u + 1 / v + 1 / w) * (1 / u + 1 / v + 1 / w) = 3 by field_simp [*]]
  let f : ℕ → ℝ | 0 => sqrt (1 / u) | 1 => sqrt (1 / v) | 2 => sqrt (1 / w) | _ => 0
  let g : ℕ → ℝ | 0 => sqrt (u) | 1 => sqrt (v) | 2 => sqrt (w) | _ => 0
  let s : Finset ℕ := {0, 1, 2}
  have h := Finset.sum_mul_sq_le_sq_mul_sq s f g
  simp [f, s] at h
  simp [g, s] at h
  (repeat rw [sq_sqrt] at h)
  rw [<- mul_le_mul_right (a := 3) (b := 3) (c:= (u + v + w) / 3 * (1 / u + 1 / v + 1 / w))]
  convert h using 1
  field_simp; norm_num;
  simp; linarith;
  all_goals positivity

/-
  The normal form of the AM-HM inequality for 3 variables
-/
theorem AM_HM_normal_left_3vars (u v w k l right : ℝ) (hu : u > 0) (hv : v > 0) (hw : w > 0) (hk : k > 0) (h : k * (u + v + w) / 9 + l ≤ right) : k * (1 / (1 / u + 1 / v + 1 / w)) + l ≤ right := by
  suffices 1 / (1 / u + 1 / v + 1 / w) ≤ (u + v + w) / 9 by nlinarith
  have h := AM_HM_3vars u v w
  field_simp [*] at h
  rw [<- mul_le_mul_right (a := 3) (b := 1 / (1 / u + 1 / v + 1 / w)) (c := (u + v + w) / 9)]
  convert h using 1
  norm_num; field_simp; linarith
  field_simp; linarith;
  positivity

theorem AM_HM_normal_right_3vars (u v w k l left : ℝ) (hu : u > 0) (hv : v > 0) (hw : w > 0) (hk : k > 0) (h : left ≤ k * (9 / (1 / u + 1 / v + 1 / w)) + l) : left ≤ k * (u + v + w) + l := by
  suffices 9 / (1 / u + 1 / v + 1 / w) ≤ u + v + w by nlinarith
  have h := AM_HM_3vars u v w
  field_simp [*] at h
  rw [<- mul_le_mul_right (a := 1 / 3) (b := 9 / (1 / u + 1 / v + 1 / w)) (c := u + v + w)]
  convert h using 1
  norm_num; field_simp; linarith
  field_simp; linarith;

/-
  The div form of the AM-HM inequality for 3 variables
-/
theorem AM_HM_div_left_3vars (u v w k l right : ℝ) (hu : u > 0) (hv : v > 0) (hw : w > 0) (hk : k > 0) (h : k * (1 / u + 1 / v + 1 / w) / 9 + l ≤ right) : k * (1 / (u + v + w)) + l ≤ right := by
  suffices 1 / (u + v + w) ≤ (1 / u + 1 / v + 1 / w) / 9 by nlinarith
  have h := AM_HM_3vars (1 / u) (1 / v) (1 / w)
  field_simp [*] at h
  rw [<- mul_le_mul_right (a := 3) (b := 1 / (u + v + w)) (c := (1 / u + 1 / v + 1 / w) / 9)]
  convert h using 1
  norm_num; field_simp;
  field_simp; linarith;
  positivity

theorem AM_HM_div_right_3vars (u v w k l left : ℝ) (hu : u > 0) (hv : v > 0) (hw : w > 0) (hk : k > 0) (h : left ≤ k * (9 / (u + v + w)) + l) : left ≤ k * (1 / u + 1 / v + 1 / w) + l := by
  suffices 9 / (u + v + w) ≤ (1 / u + 1 / v + 1 / w) by nlinarith
  have h := AM_HM_3vars (1 / u) (1 / v) (1 / w)
  field_simp [*] at h
  rw [<- mul_le_mul_right (a := 1 / 3) (b := 9 / (u + v + w)) (c := (1 / u + 1 / v + 1 / w))]
  convert h using 1
  norm_num; field_simp; linarith
  field_simp; linarith;
