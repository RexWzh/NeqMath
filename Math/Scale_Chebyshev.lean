import Mathlib

open Real

local macro_rules | `($x / $y)   => `(HDiv.hDiv ($x : ℝ) ($y : ℝ))

section Scale_Chebyshev_2vars

/-- The basic format of the cycle form Chebyshev inequality for 2 variables -/
theorem Chebyshev_cycle_2vars (u v p q : ℝ) (ha : u ≥ 0) (hb : v ≥ 0) (hc : p ≥ 0) (hd : q ≥ 0): (u ^ p + v ^ p) * (u ^ q + v ^ q) ≤ (2 * u ^ (p + q) + 2 *  v ^ (p + q)) := by
  sorry

/-- The normal form Chebyshev inequality for 2 variables
    note that normal right cannot be pattern matched, and thus ignored -/

theorem Chebyshev_cycle_left_2vars (u v p q k l right : ℝ) (ha : u ≥ 0) (hb : v ≥ 0) (hp : p ≥ 0) (hq : q ≥ 0) (hk : k ≥ 0)
  (hl : k * (2 * u ^ (p + q) + 2 * v ^ (p + q)) + l ≤ right) : k * (u ^ p + v ^ p) * (u ^ q + v ^ q) + l ≤ right := by
  suffices (u ^ p + v ^ p) * (u ^ q + v ^ q) ≤ (2 * u ^ (p + q) + 2 * v ^ (p + q)) by nlinarith
  exact Chebyshev_cycle_2vars u v p q ha hb hp hq

theorem Chebyshev_cycle_Qeq1_left_2vars (u v p k l right : ℝ) (ha : u ≥ 0) (hb : v ≥ 0) (hp : p ≥ 0) (hk : k ≥ 0)
  (hl : k * (2 * u ^ (p + 1) + 2 * v ^ (p + 1)) + l ≤ right) : k * (u ^ p + v ^ p) * (u + v) + l ≤ right := by
  suffices (u ^ p + v ^ p) * (u + v) ≤ (2 * u ^ (p + 1) + 2 * v ^ (p + 1)) by nlinarith
  have hq : (1:ℝ) ≥ 0 := by positivity
  have h := Chebyshev_cycle_2vars u v p (1:ℝ) ha hb hp hq
  norm_num at h
  exact h

end Scale_Chebyshev_2vars

section Scale_Chebyshev_3vars
/-- The basic format of the cycle form Chebyshev inequality for 3 variables -/
theorem Chebyshev_cycle_3vars (u v w p q : ℝ) (ha : u ≥ 0) (hb : v ≥ 0) (hc : w ≥ 0) (hp : p ≥ 0) (hq : q ≥ 0) : (u ^ p + v ^ p + w ^ p) * (u ^ q + v ^ q + w ^ q) ≤ (3 * u ^ (p + q) + 3 * v ^ (p + q) + 3 * w ^ (p + q)) := by
  sorry

/-- The normal form Chebyshev inequality for 3 variables
    note that normal right cannot be pattern matched, and ignored -/

theorem Chebyshev_cycle_left_3vars (u v w p q k l right : ℝ) (ha : u ≥ 0) (hb : v ≥ 0) (hc : w ≥ 0) (hp : p ≥ 0) (hq : q ≥ 0) (hk : k ≥ 0)
  (hl : k * (3 * u ^ (p + q) + 3 * v ^ (p + q) + 3 * w ^ (p + q)) + l ≤ right) : k * (u ^ p + v ^ p + w ^ p) * (u ^ q + v ^ q + w ^ q) + l ≤ right := by
  suffices (u ^ p + v ^ p + w ^ p) * (u ^ q + v ^ q + w ^ q) ≤ (3 * u ^ (p + q) + 3 * v ^ (p + q) + 3 * w ^ (p + q)) by nlinarith
  exact Chebyshev_cycle_3vars u v w p q ha hb hc hp hq

theorem Chebyshev_cycle_Qeq1_left_3vars (u v w p k l right : ℝ) (ha : u ≥ 0) (hb : v ≥ 0) (hc : w ≥ 0) (hp : p ≥ 0) (hk : k ≥ 0)
  (hl : k * (3 * u ^ (p + 1) + 3 * v ^ (p + 1) + 3 * w ^ (p + 1)) + l ≤ right) : k * (u ^ p + v ^ p + w ^ p) * (u + v + w) + l ≤ right := by
  suffices (u ^ p + v ^ p + w ^ p) * (u + v + w) ≤ (3 * u ^ (p + 1) + 3 * v ^ (p + 1) + 3 * w ^ (p + 1)) by nlinarith
  have hq : (1:ℝ) ≥ 0 := by positivity
  have h := Chebyshev_cycle_3vars u v w p (1:ℝ) ha hb hc hp hq
  norm_num at h
  exact h

/-- The div form Chebyshev inequality for 3 variables
    similarly, div left cannot be pattern matched, and ignored -/
theorem Chebyshev_div_cycle_3vars (u v w p q : ℝ) (ha : u > 0) (hb : v > 0) (hc : w > 0) (hp : p ≥ 0) (hq : q ≥ 0) :
  (3 * u ^ (p - q) + 3 * v ^ (p - q) + 3 * w ^ (p - q)) ≤ (u ^ p + v ^ p + w ^ p) * (1 / u ^ q + 1 /v ^ q + 1 / w ^ q) := by
  sorry

theorem Chebyshev_div_cycle_right_3vars (u v w p q k l left : ℝ) (ha : u > 0) (hb : v > 0) (hc : w > 0) (hp : p ≥ 0) (hq : q ≥ 0) (hk : k ≥ 0)
  (hl : left ≤ k * (3 * u ^ (p - q) + 3 * v ^ (p - q) + 3 * w ^ (p - q)) + l) : left ≤ k * (u ^ p + v ^ p + w ^ p) * (1 / u ^ q + 1 /v ^ q + 1 / w ^ q) + l := by
  suffices (3 * u ^ (p - q) + 3 * v ^ (p - q) + 3 * w ^ (p - q)) ≤ (u ^ p + v ^ p + w ^ p) * (1 / u ^ q + 1 /v ^ q + 1 / w ^ q) by nlinarith
  exact Chebyshev_div_cycle_3vars u v w p q ha hb hc hp hq

theorem Chebyshev_div_cycle_Peq1_right_3vars (u v w p k l left : ℝ) (ha : u > 0) (hb : v > 0) (hc : w > 0) (hp : p ≥ 0) (hk : k ≥ 0)
  (hl : left ≤ k * (3 * u ^ (1 - p) + 3 * v ^ (1 - p) + 3 * w ^ (1 - p)) + l) : left ≤ k * (u + v + w) * (1 / u ^ p + 1 / v ^ p + 1 / w ^ p) + l := by
  suffices (3 * u ^ (1 - p) + 3 * v ^ (1 - p) + 3 * w ^ (1 - p)) ≤ (u + v + w) * (1 / u ^ p + 1 / v ^ p + 1 / w ^ p) by nlinarith
  have hq : (1:ℝ) ≥ 0 := by positivity
  have h := Chebyshev_div_cycle_3vars u v w (1:ℝ) p ha hb hc hq hp
  simp only [rpow_one] at h
  exact h

theorem Chebyshev_div_cycle_Qeq1_right_3vars (u v w p k l left : ℝ) (ha : u > 0) (hb : v > 0) (hc : w > 0) (hp : p ≥ 0) (hk : k ≥ 0)
  (hl : left ≤ k * (3 * u ^ (p - 1) + 3 * v ^ (p - 1) + 3 * w ^ (p - 1)) + l) : left ≤ k * (u ^ p + v ^ p + w ^ p) * (1 / u + 1 / v + 1 / w) + l := by
  suffices (3 * u ^ (p - 1) + 3 * v ^ (p - 1) + 3 * w ^ (p - 1)) ≤ (u ^ p + v ^ p + w ^ p) * (1 / u + 1 / v + 1 / w) by nlinarith
  have hq : (1:ℝ) ≥ 0 := by positivity
  have h := Chebyshev_div_cycle_3vars u v w p (1:ℝ) ha hb hc hp hq
  simp only [rpow_one] at h
  exact h

end Scale_Chebyshev_3vars

section Scale_Chebyshev_4vars

/-- The basic format of the cycle form Chebyshev inequality for 4 variables -/
theorem Chebyshev_cycle_4vars (u1 u2 u3 u4 p q : ℝ) (ha : u1 ≥ 0) (hb : u2 ≥ 0) (hc : u3 ≥ 0) (hd : u4 ≥ 0) (hp : p ≥ 0) (hq : q ≥ 0) : (u1 ^ p + u2 ^ p + u3 ^ p + u4 ^ p) * (u1 ^ q + u2 ^ q + u3 ^ q + u4 ^ q) ≤ (4 * u1 ^ (p + q) + 4 * u2 ^ (p + q) + 4 * u3 ^ (p + q) + 4 * u4 ^ (p + q)) := by
  sorry

/-- The normal form Chebyshev inequality for 3 variables
    note that normal right cannot be pattern matched, and ignored -/

theorem Chebyshev_cycle_left_4vars (u1 u2 u3 u4 p q k l right : ℝ) (ha : u1 ≥ 0) (hb : u2 ≥ 0) (hc : u3 ≥ 0) (hd : u4 ≥ 0) (hp : p ≥ 0) (hq : q ≥ 0) (hk : k ≥ 0)
  (hl : k * (4 * u1 ^ (p + q) + 4 * u2 ^ (p + q) + 4 * u3 ^ (p + q) + 4 * u4 ^ (p + q)) + l ≤ right) : k * (u1 ^ p + u2 ^ p + u3 ^ p + u4 ^ p) * (u1 ^ q + u2 ^ q + u3 ^ q + u4 ^ q) + l ≤ right := by
  suffices (u1 ^ p + u2 ^ p + u3 ^ p + u4 ^ p) * (u1 ^ q + u2 ^ q + u3 ^ q + u4 ^ q) ≤ (4 * u1 ^ (p + q) + 4 * u2 ^ (p + q) + 4 * u3 ^ (p + q) + 4 * u4 ^ (p + q)) by nlinarith
  exact Chebyshev_cycle_4vars u1 u2 u3 u4 p q ha hb hc hd hp hq

theorem Chebyshev_cycle_Qeq1_left_4vars (u1 u2 u3 u4 p k l right : ℝ) (ha : u1 ≥ 0) (hb : u2 ≥ 0) (hc : u3 ≥ 0) (hd : u4 ≥ 0) (hp : p ≥ 0) (hk : k ≥ 0)
  (hl : k * (4 * u1 ^ (p + 1) + 4 * u2 ^ (p + 1) + 4 * u3 ^ (p + 1) + 4 * u4 ^ (p + 1)) + l ≤ right) : k * (u1 ^ p + u2 ^ p + u3 ^ p + u4 ^ p) * (u1 + u2 + u3 + u4) + l ≤ right := by
  suffices (u1 ^ p + u2 ^ p + u3 ^ p + u4 ^ p) * (u1 + u2 + u3 + u4) ≤ (4 * u1 ^ (p + 1) + 4 * u2 ^ (p + 1) + 4 * u3 ^ (p + 1) + 4 * u4 ^ (p + 1)) by nlinarith
  have hq : (1:ℝ) ≥ 0 := by positivity
  have h := Chebyshev_cycle_4vars u1 u2 u3 u4 p (1:ℝ) ha hb hc hd hp hq
  norm_num at h
  exact h

/-- The div form Chebyshev inequality for 4 variables
    similarly, div left cannot be pattern matched, and ignored -/
theorem Chebyshev_div_cycle_4vars (u1 u2 u3 u4 p q : ℝ) (ha : u1 > 0) (hb : u2 > 0) (hc : u3 > 0) (hd : u4 > 0) (hp : p ≥ 0) (hq : q ≥ 0) :
  (4 * u1 ^ (p - q) + 4 * u2 ^ (p - q) + 4 * u3 ^ (p - q) + 4 * u4 ^ (p - q)) ≤ (u1 ^ p + u2 ^ p + u3 ^ p + u4 ^ p) * (1 / u1 ^ q + 1 / u2 ^ q + 1 / u3 ^ q + 1 / u4 ^ q) := by
  sorry

theorem Chebyshev_div_cycle_right_4vars (u1 u2 u3 u4 p q k l left : ℝ) (ha : u1 > 0) (hb : u2 > 0) (hc : u3 > 0) (hd : u4 > 0) (hp : p ≥ 0) (hq : q ≥ 0) (hk : k ≥ 0)
  (hl : left ≤ k * (4 * u1 ^ (p - q) + 4 * u2 ^ (p - q) + 4 * u3 ^ (p - q) + 4 * u4 ^ (p - q)) + l) :
  left ≤ k * (u1 ^ p + u2 ^ p + u3 ^ p + u4 ^ p) * (1 / u1 ^ q + 1 / u2 ^ q + 1 / u3 ^ q + 1 / u4 ^ q) + l := by
  suffices (4 * u1 ^ (p - q) + 4 * u2 ^ (p - q) + 4 * u3 ^ (p - q) + 4 * u4 ^ (p - q)) ≤ (u1 ^ p + u2 ^ p + u3 ^ p + u4 ^ p) * (1 / u1 ^ q + 1 / u2 ^ q + 1 / u3 ^ q + 1 / u4 ^ q) by nlinarith
  exact Chebyshev_div_cycle_4vars u1 u2 u3 u4 p q ha hb hc hd hp hq

theorem Chebyshev_div_cycle_Peq1_right_4vars (u1 u2 u3 u4 p k l left : ℝ) (ha : u1 > 0) (hb : u2 > 0) (hc : u3 > 0) (hd : u4 > 0) (hp : p ≥ 0) (hk : k ≥ 0)
  (hl : left ≤ k * (4 * u1 ^ (1 - p) + 4 * u2 ^ (1 - p) + 4 * u3 ^ (1 - p) + 4 * u4 ^ (1 - p)) + l) :
  left ≤ k * (u1 + u2 + u3 + u4) * (1 / u1 ^ p + 1 / u2 ^ p + 1 / u3 ^ p + 1 / u4 ^ p) + l := by
  suffices (4 * u1 ^ (1 - p) + 4 * u2 ^ (1 - p) + 4 * u3 ^ (1 - p) + 4 * u4 ^ (1 - p)) ≤ (u1 + u2 + u3 + u4) * (1 / u1 ^ p + 1 / u2 ^ p + 1 / u3 ^ p + 1 / u4 ^ p) by nlinarith
  have hq : (1:ℝ) ≥ 0 := by positivity
  have h := Chebyshev_div_cycle_4vars u1 u2 u3 u4 (1:ℝ) p ha hb hc hd hq hp
  simp only [rpow_one] at h
  exact h

theorem Chebyshev_div_cycle_Qeq1_right_4vars (u1 u2 u3 u4 p k l left : ℝ) (ha : u1 > 0) (hb : u2 > 0) (hc : u3 > 0) (hd : u4 > 0) (hp : p ≥ 0) (hk : k ≥ 0)
  (hl : left ≤ k * (4 * u1 ^ (p - 1) + 4 * u2 ^ (p - 1) + 4 * u3 ^ (p - 1) + 4 * u4 ^ (p - 1)) + l) :
  left ≤ k * (u1 ^ p + u2 ^ p + u3 ^ p + u4 ^ p) * (1 / u1 + 1 / u2 + 1 / u3 + 1 / u4) + l := by
  suffices (4 * u1 ^ (p - 1) + 4 * u2 ^ (p - 1) + 4 * u3 ^ (p - 1) + 4 * u4 ^ (p - 1)) ≤ (u1 ^ p + u2 ^ p + u3 ^ p + u4 ^ p) * (1 / u1 + 1 / u2 + 1 / u3 + 1 / u4) by nlinarith
  have hq : (1:ℝ) ≥ 0 := by positivity
  have h := Chebyshev_div_cycle_4vars u1 u2 u3 u4 p (1:ℝ) ha hb hc hd hp hq
  simp only [rpow_one] at h
  exact h

end Scale_Chebyshev_4vars
