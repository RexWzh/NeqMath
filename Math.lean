-- This module serves as the root of the `Math` library.
-- Import modules here that should be built as part of the library.
import Mathlib
import Smt

import Math.NeqScales
import Math.Tactics

@[inherit_doc] prefix:max "sqrt" => Real.sqrt
