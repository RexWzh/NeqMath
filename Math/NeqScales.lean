import Math.Scale_AM_GM
import Math.Scale_Cauchy_Schwarz
import Math.Scale_Weighted_AM_GM
import Math.Scale_Power_Mean
import Math.Scale_Rearrange
import Math.Scale_Chebyshev
import Math.Scale_Muirhead
import Math.Scale_AM_HM
import Math.Scale_Jensen
import Math.Scale_Karamata
import Math.Scale_Weighted_Jensen
import Math.Scale_Titu
import Math.Scale_Schur
import Math.Scale_Holder
import Math.Scale_Vasc

-- import Math.Scale_Chebyshev
-- import Math.Scale_LOG
-- import Math.Scale_Misc
