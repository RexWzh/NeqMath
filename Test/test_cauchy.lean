import Math

#check Cauchy_Schwarz_normal_left_2vars
#check Cauchy_Schwarz_normal_right_2vars
#check Cauchy_Schwarz_sqrt_left_2vars
#check Cauchy_Schwarz_sqrt_right_2vars
#check Cauchy_Schwarz_sqrt_cycle_left_2vars
#check Cauchy_Schwarz_sqrt_cycle_right_2vars
#check Cauchy_Schwarz_normal_left_3vars
#check Cauchy_Schwarz_normal_right_3vars
#check Cauchy_Schwarz_sqrt_left_3vars
#check Cauchy_Schwarz_sqrt_right_3vars
