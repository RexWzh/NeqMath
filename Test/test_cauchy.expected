Cauchy_Schwarz_normal_left_2vars (u1 u2 v1 v2 k l right : ℝ) (h1 : k ≥ 0)
  (h2 : k * (u1 ^ 2 + u2 ^ 2) * (v1 ^ 2 + v2 ^ 2) + l ≤ right) : k * (u1 * v1 + u2 * v2) ^ 2 + l ≤ right
Cauchy_Schwarz_normal_right_2vars (u1 u2 v1 v2 k l left : ℝ) (h1 : k ≥ 0)
  (h2 : left ≤ k * (u1 * v1 + u2 * v2) ^ 2 + l) : left ≤ k * (u1 ^ 2 + u2 ^ 2) * (v1 ^ 2 + v2 ^ 2) + l
Cauchy_Schwarz_sqrt_left_2vars (u1 u2 v1 v2 k l right : ℝ) (h1 : k ≥ 0)
  (h2 : k * sqrt(u1 ^ 2 + u2 ^ 2) * sqrt(v1 ^ 2 + v2 ^ 2) + l ≤ right) : k * (u1 * v1 + u2 * v2) + l ≤ right
Cauchy_Schwarz_sqrt_right_2vars (u1 u2 v1 v2 k l left : ℝ) (h1 : k ≥ 0) (h2 : left ≤ k * (u1 * v1 + u2 * v2) + l) :
  left ≤ k * sqrt(u1 ^ 2 + u2 ^ 2) * sqrt(v1 ^ 2 + v2 ^ 2) + l
Cauchy_Schwarz_sqrt_cycle_left_2vars (u1 u2 u3 u4 u5 u6 v1 v2 v3 v4 v5 v6 k l right : ℝ) (hk : k ≥ 0)
  (h :
    k *
          (sqrt(u1 ^ 2 + u2 ^ 2) * sqrt(v1 ^ 2 + v2 ^ 2) + sqrt(u3 ^ 2 + u4 ^ 2) * sqrt(v3 ^ 2 + v4 ^ 2) +
            sqrt(u5 ^ 2 + u6 ^ 2) * sqrt(v5 ^ 2 + v6 ^ 2)) +
        l ≤
      right) :
  k * (u1 * v1 + u2 * v2 + u3 * v3 + u4 * v4 + u5 * v5 + u6 * v6) + l ≤ right
Cauchy_Schwarz_sqrt_cycle_right_2vars (u1 u2 u3 u4 u5 u6 v1 v2 v3 v4 v5 v6 k l left : ℝ) (hu1 : u1 ≥ 0) (hu2 : u2 ≥ 0)
  (hu3 : u3 ≥ 0) (hu4 : u4 ≥ 0) (hu5 : u5 ≥ 0) (hu6 : u6 ≥ 0) (hv1 : v1 ≥ 0) (hv2 : v2 ≥ 0) (hv3 : v3 ≥ 0)
  (hv4 : v4 ≥ 0) (hv5 : v5 ≥ 0) (hv6 : v6 ≥ 0) (hk : k ≥ 0)
  (h : left ≤ k * (sqrt(u1 * v1) + sqrt(u2 * v2) + sqrt(u3 * v3) + sqrt(u4 * v4) + sqrt(u5 * v5) + sqrt(u6 * v6)) + l) :
  left ≤ k * (sqrt(u1 + u2) * sqrt(v1 + v2) + sqrt(u3 + u4) * sqrt(v3 + v4) + sqrt(u5 + u6) * sqrt(v5 + v6)) + l
Cauchy_Schwarz_normal_left_3vars (u1 u2 u3 v1 v2 v3 k l right : ℝ) (h1 : k ≥ 0)
  (h2 : k * (u1 ^ 2 + u2 ^ 2 + u3 ^ 2) * (v1 ^ 2 + v2 ^ 2 + v3 ^ 2) + l ≤ right) :
  k * (u1 * v1 + u2 * v2 + u3 * v3) ^ 2 + l ≤ right
Cauchy_Schwarz_normal_right_3vars (u1 u2 u3 v1 v2 v3 k l left : ℝ) (h1 : k ≥ 0)
  (h2 : left ≤ k * (u1 * v1 + u2 * v2 + u3 * v3) ^ 2 + l) :
  left ≤ k * (u1 ^ 2 + u2 ^ 2 + u3 ^ 2) * (v1 ^ 2 + v2 ^ 2 + v3 ^ 2) + l
Cauchy_Schwarz_sqrt_left_3vars (u1 u2 u3 v1 v2 v3 k l right : ℝ) (h1 : k ≥ 0)
  (h2 : k * sqrt(u1 ^ 2 + u2 ^ 2 + u3 ^ 2) * sqrt(v1 ^ 2 + v2 ^ 2 + v3 ^ 2) + l ≤ right) :
  k * (u1 * v1 + u2 * v2 + u3 * v3) + l ≤ right
Cauchy_Schwarz_sqrt_right_3vars (u1 u2 u3 v1 v2 v3 k l left : ℝ) (h1 : k ≥ 0)
  (h2 : left ≤ k * (u1 * v1 + u2 * v2 + u3 * v3) + l) :
  left ≤ k * sqrt(u1 ^ 2 + u2 ^ 2 + u3 ^ 2) * sqrt(v1 ^ 2 + v2 ^ 2 + v3 ^ 2) + l
